### PR TITLE
[youtube:tab] Add availability metadata for playlists

### PR DESCRIPTION
**Pull request by colethedj**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Bug fix
- [ ] Improvement
- [ ] New extractor
- [X] New feature

---

### Description of your *pull request* and other information

- Adds `availability` key support for playlists.
- Adds fallback for `is_unlisted` in `YouTubeIE` (related: #501) (Note: without `microformat` we won't be able to detect if a video is public since unlisted will be None)
 
Currently only supports public, unlisted and private availability forms for playlists. I'm not aware of any other examples of the other types.

Note: as of writing, the `public` availability is only possible when the playlist is a personal public playlist and auth is given.

TODO: 
- [X] Can probably add a test of some sort

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -644,6 +644,28 @@ def _report_alerts(self, alerts, expected=True):
...
+    def _extract_badges(self, renderer: dict):
+        badges = set()
+        for badge in try_get(renderer, lambda x: x['badges'], list) or []:
+            label = try_get(badge, lambda x: x['metadataBadgeRenderer']['label'], compat_str)
+            if label:
+                badges.add(label.lower())
+        return badges
+
+    @staticmethod
+    def _join_text_entries(runs):
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:656]: This is a duplicate of the `get_text` function

**Comment by pukkandan** [comment]
>  the public availability is only possible when the playlist is a personal public playlist and auth is given.

If we don't detect the playlist as private/unlisted, we know it is public, no?

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (colethedj) [reply yt_dlp/extractor/youtube.py:656]: That's true. I think there a lot of duplicates of similar code over youtube.py, functions or not. I'll prob look into writing one general function to handle them all.

Comment (colethedj) [reply yt_dlp/extractor/youtube.py:656]: I think I'll deal with this in #505 after this PR is merged

**Comment by colethedj** [comment]
> > the public availability is only possible when the playlist is a personal public playlist and auth is given.
> 
> If we don't detect the playlist as private/unlisted, we know it is public, no?

With video availability, we are told explicitly if the video is not private and not unlisted. If we fail to extract either of those (i.e., they are None), then we can't say for certain that the video is available hence the availability won't be set to public.

Difficult thing with playlists is we don't have that information. If the playlist is public, unlisted and private are None. So, we can either assume that means the playlist is public (but also means if there is an extraction issue we may be incorrectly reporting it as public) or just have availability be None.

**Comment by pukkandan** [comment]
>  Can probably add a test of some sort

We can just create an unlisted playlist and add it to tests

**Comment by colethedj** [comment]
> > > the public availability is only possible when the playlist is a personal public playlist and auth is given.
> > 
> > 
> > If we don't detect the playlist as private/unlisted, we know it is public, no?
> 
> With video availability, we are told explicitly if the video is not private and not unlisted. If we fail to extract either of those (i.e., they are None), then we can't say for certain that the video is public hence the availability won't be set to public.
> 
> Difficult thing with playlists is we don't have that information. If the playlist is public, unlisted and private are None. So, we can either assume that means the playlist is public (but also means if there is an extraction issue we may be incorrectly reporting it as public) or just have availability be None.

Another issue with assuming it's public is this availability is extracted for all tab types. Which is fine for e.g. /videos, but it will be incorrect if extracting e.g. watch history or other areas where the availability would be private but there is no easy way of telling. So, I'll keep it None unless otherwise certain. Just means this won't be very useful for telling if a given playlist is public - only if unlisted/private. :disappointed:

**Pull request merged**